src/odemis/driver/tmcm.py

```diff
@@ -2502,3 +2553,871 @@ def _parseMessage(self, msg):
+                node.nmt.stop_node_guarding()
+            self._network.sync.stop()
+            self._network.disconnect()
+
+    def _init_axis(self, axis):
+        """
+        Initialise the given axis with "good" values for our needs
+        axis (int): axis number
+        """
+        # NOTE: Some objects can only be written when the drive is in the
+        #       SWITCHED_ON_DISABLED state (but is always readable).
+        self._node.state = SWITCH_ON_DISABLED
+        self.SetAxisParam(axis, SWITCH_PARAM, 0x3)  # Disable the limit switch inputs.
+        self.SetAxisParam(axis, PULLUP_RESISTORS, 0x1)  # Enable pull-up resistor for HOME input
+        # Encoder: 0 internal magnetic encoder, -1 (default) no encoder, -2 external encoder
+        self.SetAxisParam(axis, SENSOR_SELECTION, 0)  # Use Internal magnetic encoder
+        self.SetAxisParam(axis, POSITION_WINDOW, 0)  # enable position window 0 (target = demand position)
+        self.SetAxisParam(axis, POSITION_WINDOW_TIME, 100)  # remain in window for 100 ms before setting status bit
+
+        self._node.op_mode = HOMING_MODE
+        self.SetAxisParam(axis, HOMING_METHOD, Homing.HM_ON_POSITIVE_INDEX_PULSE)
+        self._node.state = OPERATION_ENABLED
+        self._node.op_mode = PP_MODE
+
+        self.SetAxisParam(axis, HOME_OFFSET, 0)  # set home offset (value to be set after referencing) to 0
+
+        # There is a small bug in the canopen library. By default, after entering "QUICK STOP ACTIVE" state,
+        # the state is automatically changed to "SWITCH ON DISABLED". However, when calling "QUICK STOP ACTIVE",
+        # the state setter (canopen.profiles.p402 l 445) will try to change the state to "QUICK STOP ACTIVE" in a loop
+        # until this state is reached. If the state is changed automatically from "QUICK STOP ACTIVE" to "SWITCHED ON
+        # DISABLE" before the loop condition is evaluated, it will attempt a transition from "SWITCHED ON DISABLE" to
+        # "QUICK STOP ACTIVE". This transition is not allowed, so it will raise an error.
+        # To avoid this, we change the settings to disable the automatic change to "SWITCH ON DISABLED" state
+        # and do it manually in the StopAxis() function.
+        self.SetAxisParam(axis, QUICK_STOP_OPTION, 6)  # stay in quick stop active state
+
+    # Low level functions
+    def GetVersion(self, axis=0):
+        """
+        return (int, int, int):
+             Controller ID
+             Firmware major version number
+             Firmware minor version number
+        """
+        cont = self.GetAxisParam(axis, DEVICE_NAME)
+        vmaj = self.GetAxisParam(axis, HW_VERSION)  # e.g. '1.0'
+        vmin = self.GetAxisParam(axis, IDENTITY, 3)
+        return cont, int(float(vmaj)), int(vmin)
+
+    def GetAxisParam(self, axis, param, idx=None):
+        """
+        Read the axis/parameter setting from the RAM
+        axis (0<=int<=5): axis number
+        param (0<=int<=255): parameter number
```

Comment:
Would param_nmr not be a more descriptive name?

Reply:
I use the same convention as in the previous tmcm driver (corresponding function has exact same parameters).
Usually, "axis" always refers to the axis number, "axname" or "axis_name" to the name (like 'x'). That avoids long variable names.